projects/solar/test/test_data_tx.c

```diff
@@ -0,0 +1,120 @@
+#include "data_tx.h"
+
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "can_unpack.h"
+#include "data_store.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "gpio.h"
+#include "interrupt.h"
+#include "log.h"
+#include "ms_test_helper_can.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+#include "solar_events.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+static uint16_t s_can_msg_count;
+static uint16_t s_num_callbacks;
+static uint32_t s_can_msg_values[2 * NUM_DATA_POINTS];
+static DataPoint s_can_msgs_data_points[2 * NUM_DATA_POINTS];
+static CanStorage s_can_storage = { 0 };
+static DataTxSettings s_data_tx_settings = {
+  .msgs_per_tx_iteration = 1,
+  .wait_between_tx_in_millis = 350,
+};
+
+static StatusCode prv_test_data_tx_callback_handler(const CanMessage *msg, void *context,
+                                                    CanAckStatus *ack_reply) {
+  TEST_ASSERT_EQUAL(SYSTEM_CAN_MESSAGE_SOLAR_DATA, msg->msg_id);
+  if (s_can_msg_count < 2 * NUM_DATA_POINTS) {
+    CAN_UNPACK_SOLAR_DATA(msg, (uint32_t *)&s_can_msgs_data_points[s_can_msg_count],
+                          &s_can_msg_values[s_can_msg_count]);
+    s_can_msg_count++;
+    return STATUS_CODE_OK;
+  }
+  return STATUS_CODE_RESOURCE_EXHAUSTED;
+}
+
+// Set every data point
+static void prv_test_data_tx_set_data_values(void) {
+  for (DataPoint data_point = 0; data_point < NUM_DATA_POINTS; data_point++) {
+    TEST_ASSERT_OK(data_store_set(data_point, data_point));
+  }
+}
+
+static void prv_reset_arrays(void) {
+  for (uint16_t index = 0; index < 2 * NUM_DATA_POINTS; index++) {
+    s_can_msg_values[index] = 0;
+    s_can_msgs_data_points[index] = 0;
+  }
+}
+
+static void prv_process_can_events(SoftTimerId timer_id, void *context) {
+  for (uint16_t msg = 0; msg < s_data_tx_settings.msgs_per_tx_iteration; msg++) {
+    MS_TEST_HELPER_CAN_TX(SOLAR_CAN_EVENT_TX);
+  }
+  for (uint16_t msg = 0; msg < s_data_tx_settings.msgs_per_tx_iteration; msg++) {
+    MS_TEST_HELPER_CAN_RX(SOLAR_CAN_EVENT_RX);
+  }
```

Comment:
This will break if `NUM_DATA_POINTS` isn't a multiple of `msgs_per_tx_iteration`. If this ever occurs, the last `prv_process_can_events` call will try to process more TX events than there were TXes, and this will fail because `MS_TEST_HELPER_CAN_TX` will choke on the first RX event after the TX events.

To fix this, I'd set a variable at the top of the function to `MIN(s_data_tx_settings.msgs_per_tx_iteration, NUM_DATA_POINTS - s_can_msg_count)` and use that instead so that it won't try to process more TX events than are raised. (Note that the `MIN` macro is defined in misc.h.)

Even though we're keeping `msgs_per_tx_iteration` at 1 for now pending the x86 CAN bug, we'll likely increase it in the future.